spanner/api/Spanner.Samples/InsertDataAsync.cs

```diff
@@ -55,34 +55,40 @@ public async Task InsertDataAsync(string projectId, string instanceId, string da
             new Album { SingerId = 2, AlbumId = 2, AlbumTitle = "Forever Hold your Peace" },
             new Album { SingerId = 2, AlbumId = 3, AlbumTitle = "Terrified" },
         };
+
         // Create connection to Cloud Spanner.
         using var connection = new SpannerConnection(connectionString);
         await connection.OpenAsync();
 
-        await Task.WhenAll(singers.Select(singer =>
+        await connection.RunWithRetriableTransactionAsync(async transaction =>
         {
-            // Insert rows into the Singers table.
-            using var cmd = connection.CreateInsertCommand("Singers", new SpannerParameterCollection
+            await Task.WhenAll(singers.Select(singer =>
             {
-                    { "SingerId", SpannerDbType.Int64, singer.SingerId },
-                    { "FirstName", SpannerDbType.String, singer.FirstName },
-                    { "LastName", SpannerDbType.String, singer.LastName }
-            });
-            return cmd.ExecuteNonQueryAsync();
-        }));
+                // Insert rows into the Singers table.
+                using var cmd = connection.CreateInsertCommand("Singers", new SpannerParameterCollection
+                {
+                        { "SingerId", SpannerDbType.Int64, singer.SingerId },
+                        { "FirstName", SpannerDbType.String, singer.FirstName },
+                        { "LastName", SpannerDbType.String, singer.LastName }
+                });
+                cmd.Transaction = transaction;
+                return cmd.ExecuteNonQueryAsync();
+            }));
 
-        await Task.WhenAll(albums.Select(album =>
-        {
-            // Insert rows into the Albums table.
-            using var cmd = connection.CreateInsertCommand("Albums", new SpannerParameterCollection
+            await Task.WhenAll(albums.Select(album =>
             {
-                    { "SingerId", SpannerDbType.Int64, album.SingerId },
-                    { "AlbumId", SpannerDbType.Int64, album.AlbumId },
-                    { "AlbumTitle", SpannerDbType.String,album.AlbumTitle }
-            });
-            return cmd.ExecuteNonQueryAsync();
-        }));
-        Console.WriteLine("Data inserted.");
+                // Insert rows into the Albums table.
+                using var cmd = connection.CreateInsertCommand("Albums", new SpannerParameterCollection
+                {
+                        { "SingerId", SpannerDbType.Int64, album.SingerId },
+                        { "AlbumId", SpannerDbType.Int64, album.AlbumId },
+                        { "AlbumTitle", SpannerDbType.String,album.AlbumTitle }
+                });
+                cmd.Transaction = transaction;
+                return cmd.ExecuteNonQueryAsync();
+            }));
+            Console.WriteLine("Data inserted.");
```

Comment:
nit: I think it would be better to place this line after the transaction instead of inside the transaction. Any error that is returned by the `Commit` will now be preceded by the `Data inserted` log line (i.e. at this point the data is only buffered and not really inserted).

Reply:
Great catch, thank you. I fixed it in both places.